src/dart/lib/decoder.dart

```diff
@@ -0,0 +1,894 @@
+  late final Rect viewBox;
+  
+  late final double minX, maxX, minY, maxY;
+  bool foundViewBox = false;
+  
+  final Uint8List bytes;
+  int cursor = 0;
+  
+  int get nextByte {
+    if (cursor >= bytes.length) {
+      throw FormatException('Unexpected end of file at offset $cursor.');
+    }
+    return bytes[cursor++];
+  }
+
+  static int ColorToRGBA(Color color) {
+    final int alpha = color.value >> 24;
+    final int red = (((color.value & 0x00FF0000) >> 16) * alpha) ~/ 255;
+    final int green = (((color.value & 0x0000FF00) >> 8) * alpha) ~/ 255;
+    final int blue = ((color.value & 0x000000FF) * alpha) ~/ 255;
+    return (red << 24) + (green << 16) + (blue << 8) + alpha;
+  }
+  
+  static Color RGBAToColor(int color) {
+    final int alpha = color & 0x000000FF;
+    if (alpha == 0x00) {
+      // We sometimes (for gradients) add in real color to the zero-alpha colors.
+      // These don't need to be unmultiplied further, they're already unmultiplied.
+      return Color((color & 0xFFFFFF00) >> 8);
+    }
+    assert(!isNonColor(color));
+    assert(((color & 0xFF000000) >> 24) <= alpha);
+    final int red = (((color & 0xFF000000) >> 24) * 255) ~/ alpha;
+    assert(((color & 0x00FF0000) >> 16) <= alpha);
+    final int green = (((color & 0x00FF0000) >> 16) * 255) ~/ alpha;
+    assert(((color & 0x0000FF00) >> 8) <= alpha);
+    final int blue = (((color & 0x0000FF00) >> 8) * 255) ~/ alpha;
+    return Color((alpha << 24) + (red << 16) + (green << 8) + blue);
+  }
+  
+  static const int kCREGLength = 64;
+  static const int kNREGLength = 64;
+  
+  final Uint32List customPalette = Uint32List.fromList(List<int>.filled(kCREGLength, 0x000000FF));
+  bool foundPalette = false;
+
+  final Uint32List CREG = Uint32List.fromList(List<int>.filled(kCREGLength, 0x000000FF));
+  final Float32List NREG = Float32List(kNREGLength);
+  late final double H;
+
+  int CSEL = 0;
+  int NSEL = 0;
+
+  double LOD0 = 0.0;
+  double LOD1 = double.infinity;
+
+  static bool isNonColor(int color) {
+    return (((color & 0xFF000000) >> 24) > (color & 0x000000FF))
+        || (((color & 0x00FF0000) >> 16) > (color & 0x000000FF))
+        || (((color & 0x0000FF00) >> 8) > (color & 0x000000FF));
+  }
+  static bool isGradient(int color) => (color & 0x000080FF) == 0x00008000;
+  
+  Gradient RGBAToGradient(int color, { ValueChanged<String>? onDebugDescribe }) {
+    assert(isGradient(color));
+    int NSTOPS = (color >> 24) & 0x3F;
+    final int CBASE = (color >> 16) & 0x3F;
+    final TileMode tileMode;
+    switch (color & 0x00C00000) {
+      case 0x00000000: // None
+        tileMode = TileMode.decal;
+        break;
+      case 0x00400000: // Pad
+        tileMode = TileMode.clamp;
+        break;
+      case 0x00800000: // Reflect
+        tileMode = TileMode.mirror;
+        break;
+      case 0x00C00000: // Repeat
+        tileMode = TileMode.repeated;
+        break;
+      default:
+        throw StateError('unreachable');
```

Comment:
nit: meaningful string in state error message

Reply:
Not really sure what to put here. "Compiler should be able to detect that this is unreachable"? :-)

Reply:
I was thinking just something like `Unsupported tile mode ${(color & 0x00C00000).toRadixString(16)`}`

Reply:
The only way this could get hit is if I made a typo somewhere, and if I did, then it's very likely the message would be wrong...